tools: fix the path generated to the sct file

naveenkaje opened the pull request:
### Description
The sct file path generated in the online compiler
is incorrect. Fix that by changing the correct_scatter_shebang
API to accept a FileRef object instead and use the path.

This change should go with change in online compiler that removes
the override for correct_scatter_shebang.

### Pull request type
    [x] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

Commit:
tools: fix the path generated to the sct file

The sct file path generated in the online compiler
is incorrect. Fix that by changing the correct_scatter_shebang
API to accept a FileRef object instead and use the path.

This change should go with change in online compiler that removes
the override for correct_scatter_shebang.

theotherjimmy commented on tools/export/makefile/__init__.py:
```diff
@@ -272,11 +272,10 @@ def generate(self):
             sct_file = self.resources.get_file_refs(FileType.LD_SCRIPT)[-1]
             new_script = self.toolchain.correct_scatter_shebang(
-                sct_file.path, join("..", dirname(sct_file.name)))
+                            sct_file, dirname(sct_file.name))
```
This indentation looks off. It looks like it's 12 spaces when it should be 4.

naveenkaje replied:
fixed

theotherjimmy commented on tools/export/uvision/__init__.py:
```diff
@@ -240,11 +240,11 @@ def generate(self):
+        self.resources.add_file_ref(FileType.LD_SCRIPT, sct_file_ref.name, sct_file_ref.path)
+        ctx['linker_script'] = sct_file_ref.name
+        if ctx['linker_script'] != sct_file_ref.path:
             self.generated_files.append(ctx['linker_script'])
```
I'm pretty sure that this (and the if above it) is no longer needed, as the new scatter file is in the Resources object.

bridadan replied:
@theotherjimmy Does the old linker script need to be removed from Resources?

naveenkaje replied:
lines removed as per Jimmy's comment

bridadan commented on tools/toolchains/arm.py:
```diff
@@ -284,8 +285,9 @@ def link(self, output, objects, libraries, lib_dirs, scatter_file):
         if scatter_file:
-            new_scatter = self.correct_scatter_shebang(scatter_file)
-            args.extend(["--scatter", new_scatter])
+            scatter_name = split(scatter_file)[-1]
```
I don't think this is quite right, shouldn't this be the path to the scatter_file relative to the project root @theotherjimmy ?

theotherjimmy replied:
Yes, this should be:
```suggestion
            scatter_name = dirname(scatter_file)
```

bridadan replied:
I don't think dirname is correct either, right? That wouldn't include the actual file name.
Wouldn't it be closer to `relpath`?

naveenkaje replied:
suggested changes by Jimmy have been made

bridadan reviewed (commented):
Really nice work! A few questions for @theotherjimmy as I'm still getting more familiar with this.

theotherjimmy commented on tools/toolchains/arm.py:
```diff
@@ -19,15 +19,16 @@
-from os.path import join, dirname, splitext, basename, exists, isfile
-from os import makedirs, write, remove
+from os.path import join, dirname, splitext, basename, exists, isfile, split
+from os import makedirs, write, curdir, remove
```
I don't think `curdir` is used. Could you remove this? or point me to where you use it?

naveenkaje replied:
curdir is removed

theotherjimmy commented on tools/toolchains/arm.py:
```diff
@@ -284,8 +285,9 @@ def link(self, output, objects, libraries, lib_dirs, scatter_file):
         if scatter_file:
-            new_scatter = self.correct_scatter_shebang(scatter_file)
-            args.extend(["--scatter", new_scatter])
+            scatter_name = dirname(scatter_file)
```
Oh yeah, my recommendation was really wrong here. This should probably be:
```suggestion
            scatter_name = relpath(scatter_file)
```

naveenkaje replied:
Changes made.

theotherjimmy reviewed (changes requested):
My mistake

bridadan commented on tools/export/uvision/__init__.py:
```diff
@@ -240,12 +240,10 @@ def generate(self):
-            self.generated_files.append(ctx['linker_script'])
+        sct_file_ref = self.resources.get_file_refs(FileType.LD_SCRIPT)[0]
+        sct_file_ref = self.toolchain.correct_scatter_shebang(sct_file_ref, dirname(sct_file_ref.name))
+        self.resources.add_file_ref(FileType.LD_SCRIPT, sct_file_ref.name, sct_file_ref.path)
```
Once we add the new `LD_SCRIPT` ref, don't we also need to remove the existing one? (The one referred to by `self.resources.get_file_refs(FileType.LD_SCRIPT)[0]` above)

theotherjimmy replied:
Why would we need to do that?

bridadan replied:
So we don't have multiple linker scripts in the resources object. I'm assuming that's a bad thing/nonsensical, perhaps my assumption is incorrect?

theotherjimmy replied:
There is nothing wrong with having more than one linker script here, to my knowledge.

bridadan replied:
This is the line that I was worried about: https://github.com/ARMmbed/mbed-os/blob/master/tools/resources/__init__.py#L373

It's only selecting the first option of linker scripts. I suppose in this case, this property isn't being used by the exporter and its directly setting the linker script in the project file to the new one generate, so its not an issue?

theotherjimmy replied:
Yes. That's why it's not an issue

bridadan replied:
Lovely!

bridadan committed:
Test to ensure only the last linker file is used in a project

bridadan committed:
Only store references to one linker script in a resources object.

Projects have only ever made use of one linker script per compiler.
However, the tools would store references to multiple linker scripts in
a project and select one in a non-deterministic way. Now we clear out
any existing linker script when we add a new one to a project. This way,
the last added linker script is always used.

bridadan committed:
Correct call to add_file_ref

bridadan committed:
Make files call the linker from the build dir, so correct the path
